LIPS/lip-composite_beacon_report_receiver.md

```diff
@@ -0,0 +1,158 @@
+tags: LIPs
+blockchain: ethereum
+lip: TBA
+status: Draft
+author: Eugine Mamin, Sam Kozin, Eugene Pshenichniy
+discussions-to: TBA
+created: 2022-01-14
+updated: 2022-01-16
+---
+
+# Introduce a composite oracle beacon report receiver
+
+## Simple summary
+
+To have a possibility of using two or even more Lido Oracle beacon report intercepting callbacks, we introduce a composite oracle beacon report receiver. The last one implements the `IBeaconReceiver` interface and internally holds an array of nested `IBeaconReceiver` instances and supplies iterative execution by calling `processLidoOracleReport` in a loop.
+
+## Abstract
+
+The `OrderedCallbacksArray` contract implements the aforementioned `IOrderedCallbacksArray` interface adding the access modifiers to allow storage changing calls (add/insert/remove) only originated by the `Voting` contract (having `require(msg.sender == Voting.address)` ). Another one contract (`CompositePostRebaseBeaconReceiver`) inherits `OrderedCallbacksArray` and implemens `IBeaconReportReceiver` interface while allowing to call `processLidoOracleReport` function only originated by the `LidoOracle` contract (having `require(msg.sender == LidoOracle.address)`).
+
+In the end, the `CompositePostRebaseBeaconReceiver` contract is a top-level entity implementing the current proposal.
+
+## Motivation
+
+Currently, the `LidoOracle` [contract](https://docs.lido.fi/contracts/lido-oracle) provides only one slot for a quorum report intercepting callback (see [LidoOracle.setBeaconReportReceiver](https://docs.lido.fi/contracts/lido-oracle#setbeaconreportreceiver)). We plan to occupy this slot soon with a newly developed contract implementing the [LIP-6 coverage application mechanism](https://research.lido.fi/t/lip-6-in-protocol-coverage-proposal/1468). In contrast, we have a technical vision of the future cross-chain/L2 [upgrades](https://hackmd.io/f3416OoaS2e1l2xu_bX6SA), which may require append additional callbacks propagating oracle beacon reports.
+
+## Specification
+
+We propose the following contracts interface. The code below presumes the Solidity v0.8 syntax.
+
+### `OrderedCallbacksArray` contract
+
+The contract provides callbacks storage in the form of addresses dynamic array.
+
+```solidity
+modifier onlyVoting()
+```
+Enforces the `msg.sender` to equal `voting` address set upon the contract construction.
+
+```solidity
+constructor(address _voting)
+```
+Init contract and store the provided `_voting` address to arm the `onlyVoting` modifier.
+See: `onlyVoting`.
+
+```solidity
+function addCallback(address _callback) external override onlyVoting;
+```
+Adds the provided `_callback` element at the end of the callbacks array, after its current last element.
+* Reverts if `_callback` address is zero.
+* Reverts if `msg.sender` is not equal to the stored `voting` address.
+* Reverts if the new length of the callbacks array becomes greater than the `MAX_CALLBACKS_CNT` contract-wide constant.
```

Comment:
Maybe make `uint256 public constant MAX_CALLBACKS_COUNT = 16;` an immutable? To be able to parametrize it on contract creation and to reduce storage usage a bit.

https://github.com/lidofinance/lido-dao/blob/e935ce24cf373f79bdf9d366c714a9cb7ca7e08b/contracts/0.8.9/OrderedCallbacksArray.sol#L17

Reply:
Hmm, as I know, both `immutable` and `constant` are occupy zero storage slot space.

I have found also this answer https://ethereum.stackexchange.com/a/113885 which highlights the substitutions difference.

Nevertheless, we set it to `16` just as a sanity check.

Reply:
Indeed!